src/exec.rs

```diff
@@ -0,0 +1,335 @@
+//! Handling of code execution; where interpreting actually happens
+//!
+//! This occurs in two parts: First, by generating the global scope with [`generate_global_scope`],
+//! and then by calling [`GlobalScope::exec`] to actually run the code.
+//!
+//! [`generate_global_scope`]: fn.generate_global_scope.html
+//! [`GlobalScope::exec`]: struct.GlobalScope.html#method.exec
+// NOTE: Quick-fix implementation details that *will* be changed in the future
+// are marked with "@QUICK-FIX". These should be removed in the process of
+// polishing this initial version of the interpreter.
+
+use crate::ast::{
+    BinOp, Block, Expr, ExprKind, FnArgs, FnParams, Ident, Item, ItemKind, PrefixOp, Stmt, StmtKind,
+};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::rc::Rc;
+
+pub struct GlobalScope<'a> {
+    funcs: HashMap<String, Func<'a>>,
+}
+
+struct LocalScope<'a> {
+    parent: ParentScope<'a>,
+    // Ideally these would be &'a str, but we run into borrowck issues if we try to do that.
+    variables: HashMap<String, Value>,
+}
+
+struct Func<'a> {
+    params: FnParams<'a>,
+    body: Block<'a>,
+}
+
+type Value = usize;
+
+enum ParentScope<'a> {
+    Global(&'a GlobalScope<'a>),
+    // Damn borrow checker :(
+    Local(Rc<RefCell<LocalScope<'a>>>),
+}
+
+/// Generates the global scope of functions (with their attached bodies) from the list of items
+pub fn generate_global_scope<'a>(items: Vec<Item<'a>>) -> GlobalScope<'a> {
+    let mut global = GlobalScope {
+        funcs: HashMap::new(),
+    };
+
+    for item in items.into_iter() {
+        match item.kind {
+            ItemKind::FnDecl { name, params, body } => {
+                // If we've already added a function with this name ot the global scope, we'll
+                // panic. We should have already validated that there aren't any name conflicts.
+                if global.funcs.contains_key(name.name) {
+                    panic!("function {:?} appears twice in the global scope", name);
+                }
+
+                global.funcs.insert(name.name.into(), Func { params, body });
```

Comment:
To avoid performing 2 lookups, this could be written as:
```
let conflict = global.funcs.insert(name.name.into(), Func { params, body });
if conflict.is_some() {
    panic!("function {:?} appears twice in the global scope", name);
}
```

Reply:
I would argue that it's more readable to go with the first solution. If we're truly concerned about efficiency there are many other places that could be improved first -- especially considering that this function is run once at execution initialization.

Reply:
That's fair although it seems a shame to do 2 identical look ups right next to eachother. I'll leave it up to you.